classes/that.sc

```diff
@@ -0,0 +1,329 @@
+	}
+
+	clear {
+		OSCdef(defName).free;
+		Ndef(defName).clear;
+		all[name] = nil;
+	}
+
+	input_ {|newInput|
+		input = newInput;
+		this.prUpdateDefs();
+	}
+
+	analyzerFunction_ {|newAnalyzerFunction|
+		analyzerFunction = newAnalyzerFunction;
+		this.prUpdateDefs();
+	}
+
+	prUpdateDefs {
+		this.prCreateNdef();
+		this.prCreateOscDef();
+	}
+
+	prCreateNdef {
+		analyzer = Ndef(defName, {
+			var inputChannels;
+			var analyzerResults;
+			var oscPayload;
+			var summedTriggers;
+
+			inputChannels = input.value.asArray;
+			numInputChannels = inputChannels.size;
+
+			analyzerResults = inputChannels.collect { |inputChannel, i|
+				// .value is a call on a function that returns an event with signal graphs
+				// of UGens. If it specifies the arguments (which is optional),
+				// it can make the resulting signal graphs depend on them or add extra
+				// information to the event. E.g. besides each channel, it can access
+				// all other input channels and the index of the current
+				// channel (in a multichannel analysis signal).
+				analyzerFunction.value(inputChannel, inputChannels, i);
+			};
+
+			//  prepare payload
+			analyzerResultKeys = analyzerResults.first.keys.remove(\trigger).asArray.sort;
+			// convert to [foo_ch1, foo_ch2, bar_ch1, bar_ch2]
+			oscPayload = analyzerResultKeys.collect({ |key|
+				analyzerResults.collect({ |event|
+					event[key];
+				});
+			});
+
+			if(analyzerResults.first[\trigger].isNil, {
+				Error("the analyzerFunction must return an event with a 'trigger' key/value").throw
+			});
+
+			// sum up all active triggers to perform an OR operation
+			summedTriggers = analyzerResults.collect({|r| r[\trigger]}).sum;
+
+			// same as SendReply.kr/ar but rate can be adapted dynamically
+			SendReply.perform(
+				UGen.methodSelectorForRate(summedTriggers.rate),
+				summedTriggers, // trig
+				oscChannelName, // cmdName
+				oscPayload.flat, // values we evaluated from the analyzer, index 3.. of the OSC message
+				-1, // implicitly set replyID to -1, index 2 of the OSC message
+			);
+		});
+	}
+
+	prCreateOscDef {
+		oscdef = OSCdef(defName, { |msg|
+			var values = msg[3..];
+			var event = ();
+			// msg[2] is replyID of SendReply which we set fixed to -1
+
+			if(analyzerResultKeys.size==1, {
+				// if only one key exists in we return an array and not a dict
+				event = values
+			}, {
+				// else we need to unwrap the keys according in respect to the channels
+				analyzerResultKeys.do { |key, i|
+					// the incoming message looks like [foo_ch1, foo_ch2, bar_ch1, bar_ch2]
+					// and gets transformed to (foo: [foo_ch1, foo_ch2], bar: [bar_ch1, bar_ch2])
+					var arrayOffset = i*numInputChannels;
+					event[key] = values[arrayOffset..(arrayOffset+numInputChannels-1)]
+				}
+			});
+
+			// store event also in object
+			latestValue = event;
+
+			// callback time
+			callback.value(event);
+
+		}, oscChannelName).fix;
+	}
+
+	*prCreateTrigger {|in, defaultTrigger, triggerFunction|
```

Comment:
This method is a bit opaque. We may have to change it, but fine for now.

Reply:
Its more of a convenience method caused by the [rule of 3](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))